### PR TITLE
SAML2

### PR DESCRIPTION
**Pull request by gymad**
SAML2 implementation

## Description

SAML2 Authentication added to SuiteCRM, please see in related issue: #2819 

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

### Final checklist

- [x] My code follows the code style of this project found [here](https://suitecrm.com/wiki/index.php/Coding_Standards).
- [ ] My change requires a change to the documentation.
- [x] I have read the [**How to Contribute**](https://suitecrm.com/wiki/index.php/Contributing_to_SuiteCRM) guidelines.

### CODE REVIEW

**Comment by pgorod** [comment]
@gymad maybe you want to trim down those directories before including them in the project, you're adding a bunch of stuff like SAML demos and docs that probably doesn't need to become part of SuiteCRM...

You're removing 11 files / 2600 lines
but you're adding 417 files / 75000 lines

File: modules/Users/authentication/SAML2Authenticate/SAML2Authenticate.php [hunk 1]
```diff
@@ -0,0 +1,171 @@
+<?php
+/**
+ *
+ * SugarCRM Community Edition is a customer relationship management program developed by
+ * SugarCRM, Inc. Copyright (C) 2004-2013 SugarCRM Inc.
+ *
+ * SuiteCRM is an extension to SugarCRM Community Edition developed by SalesAgility Ltd.
+ * Copyright (C) 2011 - 2016 SalesAgility Ltd.
```
Comment (samus-aran) [inline modules/Users/authentication/SAML2Authenticate/SAML2Authenticate.php]: Very minor - update license to 2017 :)

File: modules/Users/authentication/SAML2Authenticate/SAML2Authenticate.php [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+ * of this program must display Appropriate Legal Notices, as required under
+ * Section 5 of the GNU Affero General Public License version 3.
+ *
+ * In accordance with Section 7(b) of the GNU Affero General Public License version 3,
+ * these Appropriate Legal Notices must retain the display of the "Powered by
+ * SugarCRM" logo and "Supercharged by SuiteCRM" logo. If the display of the logos is not
+ * reasonably feasible for  technical reasons, the Appropriate Legal Notices must
+ * display the words  "Powered by SugarCRM" and "Supercharged by SuiteCRM".
+ */
+
```
Comment (samus-aran) [inline modules/Users/authentication/SAML2Authenticate/SAML2Authenticate.php]: Missing the EntryPoint validation

File: modules/Users/authentication/SAML2Authenticate/SAML2AuthenticateUser.php [hunk 3]
```diff
@@ -0,0 +1,91 @@
+<?php
+/**
+ *
+ * SugarCRM Community Edition is a customer relationship management program developed by
+ * SugarCRM, Inc. Copyright (C) 2004-2013 SugarCRM Inc.
+ *
+ * SuiteCRM is an extension to SugarCRM Community Edition developed by SalesAgility Ltd.
+ * Copyright (C) 2011 - 2016 SalesAgility Ltd.
```
Comment (samus-aran) [inline modules/Users/authentication/SAML2Authenticate/SAML2AuthenticateUser.php]: update license

File: modules/Users/authentication/SAML2Authenticate/SAML2AuthenticateUser.php [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+ * of this program must display Appropriate Legal Notices, as required under
+ * Section 5 of the GNU Affero General Public License version 3.
+ *
+ * In accordance with Section 7(b) of the GNU Affero General Public License version 3,
+ * these Appropriate Legal Notices must retain the display of the "Powered by
+ * SugarCRM" logo and "Supercharged by SuiteCRM" logo. If the display of the logos is not
+ * reasonably feasible for  technical reasons, the Appropriate Legal Notices must
+ * display the words  "Powered by SugarCRM" and "Supercharged by SuiteCRM".
+ */
+
```
Comment (samus-aran) [inline modules/Users/authentication/SAML2Authenticate/SAML2AuthenticateUser.php]: Missing EntryPoint validation

File: modules/Users/authentication/SAML2Authenticate/lib/onelogin/php-saml/composer.json [hunk 5]
```diff
@@ -0,0 +1,39 @@
+{
```
Comment (samus-aran) [inline modules/Users/authentication/SAML2Authenticate/lib/onelogin/php-saml/composer.json]: Remove this file for the time being - we'll move this to the main composer file in upcoming releases.

File: modules/Users/authentication/SAML2Authenticate/lib/onelogin/settings.php [hunk 6]
```diff
@@ -0,0 +1,64 @@
+<?php
+/**
+ *
+ * SugarCRM Community Edition is a customer relationship management program developed by
+ * SugarCRM, Inc. Copyright (C) 2004-2013 SugarCRM Inc.
+ *
+ * SuiteCRM is an extension to SugarCRM Community Edition developed by SalesAgility Ltd.
+ * Copyright (C) 2011 - 2016 SalesAgility Ltd.
```
Comment (samus-aran) [inline modules/Users/authentication/SAML2Authenticate/lib/onelogin/settings.php]: Update license

File: modules/Users/authentication/SAML2Authenticate/lib/onelogin/settings.php [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+ * Section 5 of the GNU Affero General Public License version 3.
+ *
+ * In accordance with Section 7(b) of the GNU Affero General Public License version 3,
+ * these Appropriate Legal Notices must retain the display of the "Powered by
+ * SugarCRM" logo and "Supercharged by SuiteCRM" logo. If the display of the logos is not
+ * reasonably feasible for  technical reasons, the Appropriate Legal Notices must
+ * display the words  "Powered by SugarCRM" and "Supercharged by SuiteCRM".
+ */
+
+$spBaseUrl = $_SERVER['REQUEST_SCHEME'].'://'.$_SERVER['SERVER_NAME'].'/';
```
Comment (samus-aran) [inline modules/Users/authentication/SAML2Authenticate/lib/onelogin/settings.php]: Get correct BasePath - issue with redirection.

**Comment by samus-aran** [comment]
Also an issue with Logging out.  If the User does not enter the Logout URL (as they maybe transititioning from SAML1 to 2) that a Fatal appears if they try and logout.

**Pull request merged**